### PR TITLE
Add passthrough Algebra instances for Alt and Ap.

### LINKED ISSUE #342: Add Algebra instance for Ap.
As @robrix points out, Ap is the identity monad transformer, so we can define a passthrough instance and allow standard effect invocation inside code like this:
```haskell
foo :: Has Blah sig m => Ap m ()
```

### PR DESCRIPTION
**Pull request by patrickt**
This enables an exceptionally fluid style of effectful programming
when working with monoidal return types.

Fixes #342.

### CODE REVIEW

File: src/Control/Algebra.hs [hunk 1]
```diff
@@ -1,4 +1,4 @@
-{-# LANGUAGE ConstraintKinds, DeriveFunctor, FlexibleInstances, FunctionalDependencies, RankNTypes, TypeOperators, UndecidableInstances #-}
+{-# LANGUAGE CPP, ConstraintKinds, DeriveFunctor, FlexibleInstances, FunctionalDependencies, RankNTypes, TypeOperators, UndecidableInstances #-}
```
Comment (robrix) [inline src/Control/Algebra.hs]: 😭

File: src/Control/Algebra.hs [hunk 2]
```diff
@@ -117,6 +118,14 @@ instance (Algebra sig m, Effect sig) => Algebra (Error e :+: sig) (Except.Except
 instance Algebra sig m => Algebra sig (Identity.IdentityT m) where
   alg = Identity.IdentityT . alg . handleCoercible
 
+#if MIN_VERSION_base(4,12,0)
+instance Algebra sig m => Algebra sig (Ap m) where
+  alg = Ap . alg . handleCoercible
```
Comment (robrix) [inline src/Control/Algebra.hs]: however,

😍

**Comment by ocharles** [comment]
> This enables an exceptionally fluid style of effectful programming
when working with monoidal return types.

Could this be demonstrated/document for us mere mortals?

**Comment by robrix** [comment]
> > This enables an exceptionally fluid style of effectful programming
> > when working with monoidal return types.
> 
> Could this be demonstrated/documented for us mere mortals?

Yes, that’s a good idea; @patrickt, would you mind adding docs to the instances?

**Comment by patrickt** [comment]
@ocharles The gist here is that instead of writing

```haskell
foo :: (Monoid a, Has … sig m) => m a
foo = mappend <$> effect1 <*> effect2
```

we can write
```haskell
foo :: (Monoid a, Has … sig m) => Ap m a
foo = effect1 <> effect2
```

The trick here is that you don’t have to apply any `Ap` new types manually, because `Ap` passes algebras through unscathed. Without it, the above would look like
```haskell
foo = Ap effect1 <> Ap effect2
```
which is less pleasing.

@robrix Excellent idea; will put out a docs patch tomorrow.

**Comment by robrix** [comment]
it turns out that a monad really _is_ a monoid in the category of endofunctors; who knew!

**Pull request merged**